### PR TITLE
Implement secondary particle bank capability

### PR DESCRIPTION
**Pull request by paulromano**
This pull request implements a secondary particle bank that can be used in the future for photon production from neutron reactions, photon transport (e.g. pair production), and subcritical multiplication. Right now, it is only being used for (n,xn) reactions where x is an integer. Note that this means for exotic reactions where the multiplicity is non-integral, the current method of multiplying the weight of the particle remains in place.

The other notable change in this pull request is that the main looping structure in the eigenvalue and fixed_source modules has been combined in a new simulation module. This should make it easier to keep them in sync in the future. The eigenvalue module now contains only eigenvalue-specific related procedures. The fixed_source module is no longer.

The last complication was that the particle track file was originally implemented based on an assumption that there is only one particle that would write to the file. I've extended the format of the file to include tracks from a primary particle and any secondaries it creates. The `openmc-track-to-vtk` script has also been modified to convert the new track file format appropriately. I've also changed the shebang line to `#!/usr/bin/env python` since Kitware finally (!) got their act together and added support for [Python 3 bindings for VTK](https://gitlab.kitware.com/vtk/vtk/merge_requests/478).

I'm planning on doing some brief performance testing and will report the results here. I don't anticipate this should have any noticeable effect since (n,2n) and (n,3n) are rather uncommon reactions. Also, since the (n,xn) physics is different, many of the test results changed. I'm planning on running through some critical benchmarks to make sure things look ok.

- [x] Performance testing results
- [x] Critical benchmark results

### CODE REVIEW

**Comment by paulromano** [comment]
I've run the Hoogenboom-Martin benchmark problem on three different achitectures (my laptop's Intel Core i5-5200U, a two-socket Sandy Bridge machine, and a two-socket Haswell machine) and I don't see any statistically significant difference in performance between the develop branch and this one, so I'm personally happy with the performance.

**Comment by walshjon** [comment]
Some testing with a system that activates the affected physics more often (e.g. fusion source) might be more telling.  It's not too surprising that the NEA benchmark doesn't show a performance hit with altered handling of (n,xn).

**Comment by paulromano** [comment]
@walshjon Good point. I'll see if I can devise a problem that maximizes (n,2n) reactions.

**Comment by paulromano** [comment]
So here's a worst-case scenario. I searched through the ENDF/B-VII.1 ACE data and found that the highest relative (n,2n) cross section is for Cr-54 at 20 MeV (where it is about 50% of the total cross section). My test problem therefore is an infinite medium of solely Cr-54 with a neutron source at 20 MeV. The develop branch gets about 10k neutrons/sec whereas on this branch it drops to 5.4k neutrons/sec. Again, this is the absolute worst it could be. In reactor problems, the impact should be negligible as I found with Hoogenboom-Martin. This is not surprising considering most (n,2n) thresholds are above 5 MeV. As you alluded to, the only "real" problem that might actually see an effect is a fusion neutronics model. I don't have any on hand so I couldn't say what the impact there would be. But I can almost guarantee that it is no worse than 2x.

**Comment by paulromano** [comment]
One other detail I neglected to mention earlier -- there is a hard limit (1000) on the number of secondary particles produced from a single primary particle. This is done so that we don't need to dynamically allocate memory from a thread during transport which can cause thread contention. For neutron-only problems, we should never come remotely close to that limit. It is more (future) designed for coupled neutron/photon and photon-only problems where many secondary particles could be produced. As a point of reference, MCNP5 uses a limit of 640.

**Comment by nelsonag** [comment]
Great addition @paulromano!.  Is this generalized enough to support easy incorporation of history splitting or is it too heavily tied to the idea of secondary particles for that?  (The only thing obvious to me at this moment is the naming scheme and the secondary particle limit of 1000, but you surely would know better).

**Comment by paulromano** [comment]
@nelsonag Yes, it should be general enough to support history splitting actually. Thanks for bringing that up!

**Comment by walshjon** [comment]
It's not exactly a stated target application, but with a 14.1 MeV D-T source at the center of a Be-reflected dirty Jezebel model, I'm seeing a 3X slowdown (Be has a pretty good (n,2n) up there and the rest is scatter).  That said, I **don't** think that this should alter the course of the PR since the functionality introduced here is definitely needed, that 3X is probably pretty close to a worst-case scenario, and the NEA full-core results are just fine.  Looks good @paulromano !

**Comment by paulromano** [comment]
@walshjon That's another good data point -- thanks! I wanted to look at performance just so we have an idea of what the effects are for typical problems. I don't know too many people (read: anyone) who are running fusion problems with OpenMC at present, so I don't think we will hear too many complaints!

**Comment by paulromano** [comment]
Here are results for the MCNP criticality validation suite, comparing the difference in k-eff between the develop branch and the secondary-bank branch. The shaded bar in the middle represents the 1sigma standard deviation of the mean of all the k-eff values.
figure_1

File: src/particle_header.F90 [hunk 1]
```diff
@@ -154,4 +161,50 @@ elemental subroutine reset_coord(this)
...
+    this % last_E         = src % E
+
+  end subroutine initialize_from_source
+
+!===============================================================================
+! CREATE_SECONDARY stores the current phase space attributes of the particle in
+! the secondary bank and increments the number of sites in the secondary bank.
+!===============================================================================
+
+  subroutine create_secondary(this, uvw, type)
```
Comment (smharper) [inline src/particle_header.F90]: Why is `uvw` an argument to this procedure?  We could just get this information from `this % coord(1) % uvw`, right?

File: src/simulation.F90 [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+  use tracking,     only: transport
+
+  implicit none
+  private
+  public :: run_simulation
+
+contains
+
+!===============================================================================
+! RUN_EIGENVALUE encompasses all the main logic where iterations are performed
```
Comment (smharper) [inline src/simulation.F90]: Typo: `RUN_EIGENVALUE` should be `RUN_SIMULATION`

File: src/particle_header.F90 [hunk 3]
```diff
@@ -154,4 +161,50 @@ elemental subroutine reset_coord(this)
...
+    integer :: n
+
+    n = this % n_secondary + 1
+    this % secondary_bank(n) % wgt    = this % wgt
+    this % secondary_bank(n) % xyz(:) = this % coord(1) % xyz
+    this % secondary_bank(n) % uvw(:) = uvw
+    this % secondary_bank(n) % E      = this % E
+    this % n_secondary = n
+
+  end subroutine create_secondary
```
Comment (smharper) [inline src/particle_header.F90]: This procedure should probably have some check like `if n > MAX_SECONDARY then call fatal_error()`

**Comment by smharper** [comment]
Thanks for removing some of the redundancy between fixed source and eigenvalue!  I think that's a non-trivial improvement.

I'm curious, why don't we put secondary particles in the same bank that use for fission source sites?  It doesn't seem to me that there is anything special about secondary particles born in (n,xn) reactions versus fission reactions.

One small detail that I don't like about this change is that `transport()` now actually transports more than one particle.  I think the code would read more clearly if `run_simulation` kept up with the secondary bank, and then called `transport` on each of the secondaries, individually.

File: src/particle_header.F90 (hunk 1 above)
Comment (paulromano) [inline src/particle_header.F90]: The angle of the secondary particle is not necessarily the same as the angle of the particle. For example, in photon transport when pair production occurs, two 0.511 MeV photons are emitted isotropically.

File: src/particle_header.F90 (hunk 3 above)
Comment (paulromano) [inline src/particle_header.F90]: Good point -- I'll add that

**Comment by paulromano** [comment]
@smharper There is something special about fission neutrons as opposed to (n,xn) neutrons -- that is how the classic k-eigenvalue/eigenvector problem is defined. If you grouped (n,xn) neutrons into the production term, you change both the eigenvalue and the eigenvector (albeit in a very small way). Red Cullen has a nice [report](http://home.comcast.net/~redcullen1/Papers/K-eff/K-eff.pdf) about the definition of k-eff. If nothing else, his reports are amusing for their unusual candor.

Regarding the use of `transport` -- to me I find it cleaner that `transport` says "transport this source particle and any of its children" rather than "transport only this particle but not its children". In the latter case, the loop size is non-deterministic, which means it becomes very difficult to divide work for threading. Maybe its just a wording issue, i.e. "transport" might not be the best name for the routine.

File: src/particle_header.F90 (hunk 1 above)
Comment (smharper) [inline src/particle_header.F90]: Okay, I see.  I imagine we'll have to add an argument for energy at some point, right?  You can leave it for a later PR though, if you like.

**Comment by smharper** [comment]
Thanks for the link.  I found it to be informative.

I guess the new `transport` routine makes sense if I think of OpenMC only conserving _fission_ neutrons from generation to generation rather than conserving _all_ neutrons.

Do you think this PR is ready to be merged?  I'm happy with it.

**Comment by paulromano** [comment]
I think it's ready to go. @walshjon Do you have any requested changes before merging?

File: src/initialize.F90 [hunk 4]
```diff
@@ -141,13 +140,9 @@ subroutine initialize_run()
...
       call calculate_work()
 
-      ! Allocate banks and create source particles -- for a fixed source
-      ! calculation, the external source distribution is sampled during the
-      ! run, not at initialization
-      if (run_mode == MODE_EIGENVALUE) then
-        call allocate_banks()
-        if (.not. restart_run) call initialize_source()
-      end if
+      ! Allocate source bank, and for eigenvalu simulations also allocate the
```
Comment (walshjon) [inline src/initialize.F90]: eigenvalue

**Comment by walshjon** [comment]
Nothing else from me on this one.  This will help enable many nice features, @paulromano .

**Comment by paulromano** [comment]
@smharper @walshjon All comments are resolved if one of you wants to hit the merge button.

**Pull request merged**